instrumentation/jetty-httpclient/jetty-httpclient-12.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/httpclient/v12_0/JettyHttpClient12Instrumentation.java

```diff
@@ -0,0 +1,124 @@
+ */
+
+package io.opentelemetry.javaagent.instrumentation.jetty.httpclient.v12_0;
+
+import static io.opentelemetry.javaagent.instrumentation.jetty.httpclient.v12_0.JettyHttpClientSingletons.JETTY_CLIENT_CONTEXT_KEY;
+import static io.opentelemetry.javaagent.instrumentation.jetty.httpclient.v12_0.JettyHttpClientSingletons.instrumenter;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.util.VirtualField;
+import io.opentelemetry.instrumentation.jetty.httpclient.v12_0.internal.JettyHttpClient12TracingInterceptor;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.eclipse.jetty.client.Request;
+import org.eclipse.jetty.client.transport.HttpRequest;
+
+public class JettyHttpClient12Instrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("org.eclipse.jetty.client.transport.HttpRequest");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(named("send"))
+            .and(takesArgument(0, named("org.eclipse.jetty.client.Response$CompleteListener"))),
+        JettyHttpClient12Instrumentation.class.getName() + "$JettyHttpClient12SendAdvice");
+    // For request listeners
+    transformer.applyAdviceToMethod(
+        isMethod().and(nameContains("notify")),
+        JettyHttpClient12Instrumentation.class.getName() + "$JettyHttpClient12NotifyAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class JettyHttpClient12SendAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnterSend(
+        @Advice.This HttpRequest request,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      // start span
+      Context parentContext = Context.current();
+      JettyHttpClient12TracingInterceptor interceptor =
+          new JettyHttpClient12TracingInterceptor(parentContext, instrumenter());
+      interceptor.attachToRequest(request);
+      context = interceptor.getContext();
+      if (context == null) {
+        return;
+      }
+      // set context for responseListeners
+      VirtualField<Request, Context> virtualField = VirtualField.find(Request.class, Context.class);
+      virtualField.set(request, parentContext);
+      request.attribute(JETTY_CLIENT_CONTEXT_KEY, parentContext);
```

Comment:
why do you need to use both `VirtualField` and set the attribute, won't just setting the attribute be enough?

Reply:
Sorry, I forgot to get request by calling `response.getRequest()`, so I use VirtualField...